--- sympy/simplify/simplify.py	2008-11-19 16:12:13.000000000 -0500
+++ sympy/simplify/simplify.py	2009-01-27 09:20:36.000000000 -0500
@@ -784,6 +784,7 @@
         )
 
         expr = ret
+        
         for pattern, result, ex in artifacts:
             # Substitute a new wild that excludes some function(s)
             # to help influence a better match. This is because
@@ -792,12 +793,22 @@
             pattern = pattern.subs(a, a_t)
             result = result.subs(a, a_t)
 
-            m = expr.match(pattern)
+            if expr.is_number:
+                continue
+            try:
+                m = expr.match(pattern)
+            except (TypeError):
+                break
             while m is not None:
                 if m[a_t] == 0 or -m[a_t] in m[c].args or m[a_t] + m[c] == 0:
                     break
                 expr = result.subs(m)
-                m = expr.match(pattern)
+                if expr.is_number:
+                    continue
+                try:
+                    m = expr.match(pattern)
+                except (TypeError):
+                    break
 
         return expr
     return expr
